File: Algorithms/61. Rotate List/golang/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	var oldTail *ListNode = head
	var n int = 1
	for oldTail.Next != nil {
		oldTail = oldTail.Next
		n++
	}
	oldTail.Next = head

	var newTail *ListNode = head
	for i := 0; i < n-k%n-1; i++ {
		newTail = newTail.Next
	}
	newHead := newTail.Next
	newTail.Next = nil
	return newHead
}

File: Algorithms/61. Rotate List/swift/solution.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func rotateRight(_ head: ListNode?, _ k: Int) -> ListNode? {
        if head == nil || head?.next == nil {
            return head
        }

        var oldTail: ListNode? = head
        var n: Int = 1
        while oldTail?.next != nil {
            oldTail = oldTail?.next
            n += 1
        }
        oldTail?.next = head

        var newTail: ListNode? = head
        for _ in 0..<(n - k % n - 1) {
            newTail = newTail?.next
        }
        let newHead: ListNode? = newTail?.next
        newTail?.next = nil
        return newHead
    }
}